File: merkle-trie/src/main/java/net/consensys/cava/trie/GetVisitor.java
package net.consensys.cava.trie;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.concurrent.AsyncResult;

final class GetVisitor<V> implements NodeVisitor<V> {
  private final AsyncResult<Node<V>> nullNodeResult = AsyncResult.completed(NullNode.instance());

  @Override
  public AsyncResult<Node<V>> visit(ExtensionNode<V> extensionNode, Bytes path) {
    return extensionNode.path().then(extensionPath -> {
      int commonPathLength = extensionPath.commonPrefixLength(path);
      assert commonPathLength < path.size() : "Visiting path doesn't end with a non-matching terminator";

      if (commonPathLength < extensionPath.size()) {
        // path diverges before the end of the extension, so it cannot match
        return AsyncResult.completed(NullNode.instance());
      }

      return extensionNode.child().then(child -> child.accept(this, path.slice(commonPathLength)));
    });
  }

  @Override
  public AsyncResult<Node<V>> visit(BranchNode<V> branchNode, Bytes path) {
    assert path.size() > 0 : "Visiting path doesn't end with a non-matching terminator";

    byte childIndex = path.get(0);
    if (childIndex == CompactEncoding.LEAF_TERMINATOR) {
      return AsyncResult.completed(branchNode);
    }

    return branchNode.child(childIndex).then(child -> child.accept(this, path.slice(1)));
  }

  @Override
  public AsyncResult<Node<V>> visit(LeafNode<V> leafNode, Bytes path) {
    return leafNode.path().then(leafPath -> {
      if (leafPath.commonPrefixLength(path) != leafPath.size()) {
        return nullNodeResult;
      }
      return AsyncResult.completed(leafNode);
    });
  }

  @Override
  public AsyncResult<Node<V>> visit(NullNode<V> nullNode, Bytes path) {
    return nullNodeResult;
  }
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/GetVisitor.kt
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes
import net.consensys.cava.trie.CompactEncoding

internal class GetVisitor<V> : NodeVisitor<V> {

  override suspend fun visit(extensionNode: ExtensionNode<V>, path: Bytes): Node<V> {
    val extensionPath = extensionNode.path()
    val commonPathLength = extensionPath.commonPrefixLength(path)
    assert(commonPathLength < path.size()) { "Visiting path doesn't end with a non-matching terminator" }

    if (commonPathLength < extensionPath.size()) {
      // path diverges before the end of the extension, so it cannot match
      return NullNode.instance()
    }

    return extensionNode.child().accept(this, path.slice(commonPathLength))
  }

  override suspend fun visit(branchNode: BranchNode<V>, path: Bytes): Node<V> {
    assert(path.size() > 0) { "Visiting path doesn't end with a non-matching terminator" }

    val childIndex = path.get(0)
    if (childIndex == CompactEncoding.LEAF_TERMINATOR) {
      return branchNode
    }

    return branchNode.child(childIndex).accept(this, path.slice(1))
  }

  override suspend fun visit(leafNode: LeafNode<V>, path: Bytes): Node<V> {
    val leafPath = leafNode.path()

    if (leafPath.commonPrefixLength(path) != leafPath.size()) {
      return NullNode.instance()
    }

    return leafNode
  }

  override suspend fun visit(nullNode: NullNode<V>, path: Bytes): Node<V> = NullNode.instance()
}
